src/helpers/subscribe.ts

```diff
@@ -0,0 +1,208 @@
+   * changes to.
+   */
+  async addTableListener(tableName: string): Promise<EventEmitter> {
+    if (tableName == null) {
+      throw new Error("table name is required to add listener");
+    }
+
+    const tableIdentifier = await getTableIdentifier(tableName);
+    const listenerId = `_${tableIdentifier.chainId}_${tableIdentifier.tableId}`;
+    if (this.listeners[listenerId]?.emitter != null) {
+      return this.listeners[listenerId].emitter;
+    }
+
+    const emitter = new EventEmitter();
+
+    // If not already listening to the contract we will start listening now,
+    // if already listening we will start tracking the new emitter.
+    await this._ensureListening(listenerId, emitter);
+
+    this.listeners[listenerId] = {
+      ...tableIdentifier,
+      emitter,
+    };
+
+    return emitter;
+  }
+
+  removeTableListener(params: TableIdentifier): void {
+    if (params == null) {
+      throw new Error("must provide chainId and tableId to remove a listener");
+    }
+
+    const listenerId = `_${params.chainId}_${params.tableId}`;
+    if (this.listeners[listenerId] == null) {
+      throw new Error("cannot remove listener that does not exist");
+    }
+
+    const emitter = this.listeners[listenerId].emitter;
+    emitter.removeAllListeners();
+
+    // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+    delete this.listeners[listenerId];
+  }
+
+  // remove all listeners and stop listening to the contract
+  removeAllTableListeners(): void {
+    // for (const contract in this.contracts) {
+    // TODO: remove contract listeners
+    // contract.off(); remove each of `contractEvents``
+    // }
+
+    for (const listener in this.listeners) {
+      const l = this.listeners[listener];
+      this.removeTableListener({
+        chainId: l.chainId,
+        tableId: l.tableId,
+      });
+    }
+  }
+
+  async _ensureListening(
+    listenerId: string,
+    emitter: EventEmitter
+  ): Promise<TablelandTables> {
+    const { chainId, tableId } = await getTableIdentifier(listenerId);
+
+    const contract = await this._getContract(chainId);
+    this._attachEmitter(contract, emitter, { tableId, chainId });
+
+    return contract;
+  }
+
+  async _getContract(chainId: number): Promise<TablelandTables> {
+    if (this.contracts[chainId] != null) return this.contracts[chainId];
+    if (this.config.signer == null) {
+      throw new Error("signer information is required to get contract");
+    }
+
+    const { contract } = await getContractAndOverrides(
+      this.config.signer,
+      chainId
+    );
+    this.contracts[chainId] = contract;
+
+    return contract;
+  }
+
+  _attachEmitter(
+    contract: TablelandTables,
+    emitter: EventEmitter,
+    tableIdentifier: TableIdentifier
+  ): void {
+    const { tableId, chainId } = tableIdentifier;
+    for (const key in contractEvents) {
+      const eve = contractEvents[key];
+      // TODO: might have to put the listener function in memory so we can remove it.
```

Comment:
Should we handle these TODOs?

Reply:
Definitely!  I've got a big set of changes that should handle these.  Should have it pushed in a few hours.